### ISSUE TITLE
Does this library make sense?

### ISSUE THREAD

**Issue by hannahhoward** [ipfs/go-graphsync#14]
When I joined the Graphsync team I feel like I did not understand a lot about the problem we’re trying to solve. I want to write down what I think I know understand to see if it is correct.

As I now understand it, the IPLD team is working on a library to handle linked data that could be in any format, traversed in arbitrary ways with links between data sometimes crossing block boundaries and sometimes not, assuming the data even uses blocks as the underlying storage. In other words, it’s an incredibly generalized concept.

Filecoin has a very specific use case: it needs to sync Blockchains quickly without multiple network round trips. And a blockchain is possibly the simplest possible concrete form of linked data, where each node is one block with one link to its parent.

I think I now understand why the Graphsync wire protocol we’re building isn’t the ideal IPLD replication protocol, because it primarily deals with blocks instead of nodes. (aside from separate discussions about whether it’s too complicated to support multiple requests and responses). 

To implement the current protocol, this library’s interactions with the selector traversal functions are going to be somewhat upside down. Because the protocol is dealing mostly in blocks, most of the work in this library is going to be accomplished by overriding the LinkLoader function (or more specifically the underlying RawLoader function) used by IPLD:

1. In the responder, when IPLD’s selector traversal functions request a block, I will load it from a local datastore (using some kind of function supplied to me by filecoin I assume) , and then I’m going to also write them out to the network to transmit to the requestor.

2. In the requestor, I’m going to supply a loader function that load blocks from the network rather than a local data store, as they come in, essentially supplying an IOReader that blocks on first Read till I get the response from the network.

It’s not entirely clear how I will use the Visit functions, except to transmit data back to Filecoin in the requestor, assuming it wants nodes and not blocks.

Anyway, I notice that:

1. Filecoin has a immediate specific need

2. The IPLD team is working on a somewhat related problem which is much more complex and generalized.

3. Trying to use IPLD’s work to solve Filecoin’s need by implementing a somewhat generalized but still less than ideal replication protocol called Graphsync, which itself will require a lot of shoe-horning, may not be the most expedient or clean way to get to solving Filecoin’s need.

So basically, I just want to make sure the current course makes sense and we want to continue on it.

I feel capable of writing this Graphsync library, I just think it’s gonna end up looking pretty weird.

**Comment by Stebalien** [comment]
I think I'm still missing way too much context the current go-ipld-prime and selector interfaces to understand the issue at hand.

Graphsync deals with blocks because blocks are the unit of serialization in IPLD. You're analysis of how we should implement this (override the LinkLoader) sounds correct (although the selector should be able to continue in the presence of not-found errors).

> It’s not entirely clear how I will use the Visit functions, except to transmit data back to Filecoin in the requestor, assuming it wants nodes and not blocks.

I believe Filecoin will be passing a no-op visitor for now. Filecoin wants graphsync so it can download some sub-graph all at once (sync the chain) but it already has a bunch of graph traversal logic that works just fine, assuming the chain is already local.

---

In the future, I'd like to be able to write things like:

```go
var obj SomeStruct
ipld.Load("/path/to/object", &obj)
```

In this case, `Load` should:

1. Create a selector describing the sub-graph we need from `/path/to/object` and the fields in `obj` (using reflection).
2. Pass this selector off to graphsync along with a visitor that fills in `obj` with the visited nodes.

A more complex example would be walking a chain and not caring about object boundaries:

```go
// Custom struct. We should use reflection to fill this with a fragment of the graph.
type CustomStruct struct {
  Parent cid.Cid
  Message string
  Data cid.Cid
  Height int
}

// tells the walker how to get from one "object" to the next.
nextSelector := "{where (.height > 10) -> select parent }" // fake selector.

// `Walk` should combine the root, the `nextSelector`, and a selector derived from
// `CustomStruct` to create a selector that describes the entire graph fragment the
// traversal will need.
//
// The walker will then _internally_ use the visitor to fill in instances of `CustomStruct`.
ipld.Walk(rootCid, nextSelector, func(x CustomStruct) {
    // do something...
})
```

Ideally, Filecoin would use something like this to be block-boundary agnostic. However, Filecoin can always switch over to a generalized walker like this in the future so it's not immediately important.

**Comment by warpfork** [comment]
I'm feeling basically +1 to all of stebaliens's comment.

The use of linkloader to do all these interesting duties -- on one side, both loading data from disk and hanging onto that same buffer to send over the wire momentarily -- and on the other side, receiving data from the wire and (tentatively?) writing it to disk as we go and validate the walk -- was not necessarily my first intuition either, fwiw.  But after sleeping on it a while, I think it's going to turn out really elegant and it seems overall appropriate.  There's a lot of things that could fit in that linkloader interface which would also be useful and powerful: other kinds of memory caching strategy or preloaders, etc.  Heck, we could wrap entire invokations of graphsync _inside_ other linkloaders to use it as the prefetcher logic in a higher level traversal.  That's... relatively far future at the moment, but the point is, yeah, having nontrivial logic in a linkloader seems totally reasonable to me.

**Comment by hannahhoward** [comment]
Ok! Thanks for bearing with me. I feel much better knowing we have shared understanding about how I am going to proceed forward. I'm going to go ahead and close this issue, though feel to comment further.
